bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/internal/scheduler/SchedulerImpl.java

```diff
@@ -0,0 +1,268 @@
+/**
+ * Copyright (c) 2014,2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.smarthome.core.internal.scheduler;
+
+import java.time.Clock;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.ZonedDateTime;
+import java.time.temporal.Temporal;
+import java.util.List;
+import java.util.concurrent.Callable;
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.scheduler.ScheduledCompletableFuture;
+import org.eclipse.smarthome.core.scheduler.Scheduler;
+import org.eclipse.smarthome.core.scheduler.SchedulerRunnable;
+import org.eclipse.smarthome.core.scheduler.SchedulerTemporalAdjuster;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Implementation of the {@link Scheduler}.
+ *
+ * @author Peter Kriens - initial contribution and API
+ * @author Simon Kaufmann - ported to CompletableFuture
+ * @author Hilbrand Bouwkamp - improved implementation and moved cron and periodic to own implementations.
+ */
+@Component(service = SchedulerImpl.class, immediate = true)
+@NonNullByDefault
+public class SchedulerImpl implements Scheduler {
+
+    private static final int THREAD_POOL_SIZE = 10;
+    private static final long THREAD_TIMEOUT_SECONDS = 65L;
+
+    private final Logger logger = LoggerFactory.getLogger(SchedulerImpl.class);
+
+    private final Clock clock = Clock.systemDefaultZone();
+    private @NonNullByDefault({}) ScheduledExecutorService executor;
+
+    @Activate
+    void activate() {
+        executor = Executors.newScheduledThreadPool(THREAD_POOL_SIZE);
```

Comment:
It would actually be nice to use the `ThreadPoolManager` here as it names the threads (which makes it better for debugging) and also allows to configure their size through config admin.

Reply:
This was one of the parts I was wondering if it should stay this way.

I have 2 questions:
1. I've moved initialization to a final field instead in the activation part. Is that the way to do it? I looked through other uses and all seem to do it that way (even static final initialization)
1. Looking at the larger architecture. Would it make sense to replace most (all?) usages of the ThreadScheduler (specifically those using repeated scheduling) with using the Scheduler implemented in this PR? Besides getting the scheduler as a service there it will all be in one large threadpool (Is it needed to have all these different threadpools? It allows for very fine tuning, but is that really needed) Also `scheduleWithFixedDelay` could be implemented as a `PeriodicScheduler` and operate at a fixed rate so be more reliable in terms of when things are executed. Actually my implementation also returns a `ScheduledFuture` so this would makes adoption easier.

Reply:
Sorry for my late reply, christmas came in my way ;-)

1. Yes, this should be alright. It is a bit awkward that the `ThreadPoolManager` does not support a proper lifecycle management, but under this constraint, a final field is just fine.
2. I would leave this out of scope. The `scheduleWithFixedDelay` is a very simple job scheduling that is used all over the place and working well - and yes, we need the different kind of thread pools as some are for the core framework, while others are used for 3rd party code and we must not mix those up. Also note that our coding guidelines specifically say that "withFixedDelay" should be used instead of "atFixedRate" - and I guess the `PeriodicScheduler` would rather result in the latter.